### ISSUE TITLE
feat: i/o

### ISSUE THREAD

**Issue by ndrewh** [ndrewh/pyda#17]
p.r should be a pwntools tube.

`p.r.sendline(....)`, etc

The tricky part is probably avoiding reading prints performed by the tool itself. Should we automatically redirect program stdout to a pipe, and leave tool stdout using regular stdout?

More precisely:
```c
int orig_in = dup(0);
int orig_out = dup(1);
int orig_err = dup(2);

int pipe1[2], pipe2[2], pipe3[2];
pipe(pipe1); pipe(pipe2); pipe(pipe3);

dup2(pipe1[0], 0)
dup2(pipe2[1], 1)
dup2(pipe3[1], 2)

stdin = fdopen(orig_in, "r");
stdout = fdopen(orig_out, "w");
stderr = fdopen(orig_err, "w");
// also need to modify dynamorio printing functions
// as they use raw fd 0/1/2
```

Then the problem becomes: who is reading from the pipe, and when?

This is a bit messy (see #13) but my guess is the right pattern is to make all recv calls implicit `pyda_yield`s. `.recvuntil` would do a non-blocking yield before entering a blocking `recv`.

recv* calls would only be allowed in two contexts:
- the top-level script (e.g. in lieu of `p.run()`).
- thread entry hooks (e.g. this is in lieu of the implicit yield that currently happens at the end of the thread entry hook)

This paradigm is quite different from the current one. Python code and target code will now be 'running' at the same time! This can cause weird races when hooks get called, but we still have the GIL so should be fine.

**Comment by ndrewh** [comment]
I'm thinking we keep the current behavior by default, and enable this with process(capture_output=True). All recv calls will assert that this options was set.

**Comment by ndrewh** [comment]
An alternative to having recv's yield _forever_ would be to have them yield just until the (next?) `write` syscall performed by the target. Then if they would still block they just yield again.

The benefit of this would be that recv_until() would leave the application blocked after it is completed (perhaps this is what users would expect?). However, I'm not sure it would be reliable, since we would basically have to rely on `read/recv` always providing data immediately on the other side of the pipe after the write syscall (maybe this is well-defined? but afaik pipes are not a concurrency primitive).

**Comment by ndrewh** [comment]
After implementing this, we can remove the [stdout close hack](https://github.com/ndrewh/pyda/blob/80178785877a4d276680714254431a3c9722490e/lib/pyda/process.py#L15-L23).

**Comment by ndrewh** [comment]
fixed in 720deee9b092a5e1966e61c62ad9ca78f201c1d3
